tests/ssl-opt.sh

```diff
@@ -1628,13 +1628,18 @@ run_test() {
         requires_config_enabled MBEDTLS_SSL_PROTO_DTLS
     fi
 
-    # If the client or server requires certain features that can be detected
-    # from their command-line arguments, check that they're enabled.
-    TLS_VERSION=$(get_tls_version "$SRV_CMD" "$CLI_CMD")
-
     # Check if we are trying to use an external tool wich does not support ECDH
     EXT_WO_ECDH=$(use_ext_tool_without_ecdh_support "$SRV_CMD" "$CLI_CMD")
 
+    # Guess the TLS version which is going to be used
```

Comment:
As part of the review of #8526, I am investigating in the configuration of all.sh component test_tls13_only_psk + MBEDTLS_SSL_RECORD_SIZE_LIMIT why "Record Size Limit: TLS 1.3: Server-side parsing and debug output" is run but "Record Size Limit: TLS 1.3: Client-side parsing and debug output" is skipped. I ended up here and I do not understand this change. With this commit, given the checks in `use_ext_tool_without_ecdh_support`, it basically means that as soon as we use GnuTLS in one test TLS_VERSION is going to be TLS12 and thus that the dependencies checks for TLS 1.3 tests below using "detect_required_features" are going to be wrong.
@valeriosetti @mpg @paul-elliott-arm @tom-daubney-arm what am I missing here?

Reply:
Question: I checked out at #8526 (current status) and then I added `scripts/config.py set   MBEDTLS_SSL_RECORD_SIZE_LIMIT` to `component_test_tls13_only_psk`. However in this configuration I get both `"Record Size Limit: TLS 1.3: (Client|Server)-side parsing and debug output"` being skipped. I'm running this in Docker with Ubuntu 16.04 so GnuTLS versions should be fine. Am I missing something?

A part from this, if in the same Docker container I issue the command `$GNUTLS_CLI -l` to list all available ciphers in GnuTLS, I don't see any TLS1.3 listed there. TBH I don't remember exactly the reason for which I added the check you mentioned, but this sounds like a possible reason. Wdyt?

Reply:
I don't think you're missing anything and I can no longer understand why things are the way they are, so either I missed an issue in reviewing this, or we really should have added more comments - which would also be something I missed when reviewing this...

Specifically:
- I think `use_ext_tool_without_ecdh_support` is misleading since is document talks about _static_ ECDH while the name suggests it's about all forms of ECDH. (I know ephemeral is called ECDHE in TLS, but still, "ECDH" does not clearly convey "static" to me.) Actually a better name for the current implementation would be `peer_is_not_mbedtls()` or something. (Similarly for the variable holding the result.)
- I have no idea why we always assume TLS 1.2 when using OpenSSL or GnuTLS as the peer. This might be justified in some particular cases but does not seem correct in general.

Reply:
Thanks for looking at this. #8526 has evolved since my previous comment thus now (as of 55dfa2c06) if you remove the dependency "requires_config_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED" you restore the issue. I've just checked it again and the "Server-side" test run and fail (as expected because we need the ephemeral key exchange mode and it is not enabled) while the "Client-side" is skipped where I would expect it to run and fail as well.   
Otherwise, the tests use G_NEXT_CLI/SRV. On my machine, the tests are run with GnuTLS 3.7.2 (version that supports TLS 1.3) and in tests/docker/bionic/Dockerfile I can see:
```
ENV GNUTLS_NEXT_CLI=/usr/local/gnutls-3.7.2/bin/gnutls-cli                      
ENV GNUTLS_NEXT_SERV=/usr/local/gnutls-3.7.2/bin/gnutls-serv
```
Thus you should be able to reproduce in Docker.

Reply:
It's quite possible that I made some mistake, perhaps I wrote something for TLS ≤1.2 and didn't realize that it didn't work the way I intended for 1.3?

I'm not sure what `use_ext_tool_without_ecdh_support` is about. AFAIR there are some versions of OpenSSL/GnuTLS/both that only support ECDHE and not ECDH, but that wouldn't affect TLS 1.3, only ≤1.2.

Reply:
>I have no idea why we always assume TLS 1.2 when using OpenSSL or GnuTLS as the peer. This might be justified in some particular cases but does not seem correct in general.

What I suspect now, but can't be sure, is that this was written (and reviewed) in a context where we were focusing on things failing with static-ECDH tests, and we simply forgot the more general context.

Now, I think the only way for a test case to require static-ECDH is if an `ECDH_xxx` ciphersuite is forced (otherwise ephemeral ECDH will be available and preferred). So if we somehow need special handling of static-ECDH I think we should look at the `force_ciphersuite` command line option to determine if we're using static-ECDH or not.

But I'm not sure we should need special handling of static-ECDH: presumably the test cases that force a static-ECDH ciphersuite don't try to use a peer that doesn't support static-ECDH.

Also, when guessing which TLS version will be used, we should have a look at `force_ciphersuite` too: if the ciphersuite name contains `WITH` then it's TLS 1.2 otherwise TLS 1.3.

Reply:
@ronald-cron-arm I think you should feel free to modify this in any way that looks logical to you, as long as it keeps the CI passing and doesn't result in more test being skipped in the default or full config.

That is, I don't think there's any clever reason that you might be missing for why things are this way. We probably just made a mistake and introduced a bug while trying to fix another one. This evolved quite organically (mostly in response to CI failures) so it's quite likely that we just lost track of the big picture at some point.

I guess it wouldn't hurt to take a step back and consider how we'd like all this auto-detection to work, and possibly restructure it a bit. (Not sure if now's the right time though.)

Reply:
Filed as https://github.com/Mbed-TLS/mbedtls/issues/9560.

(After spending hours trying to understand what was going on :frowning_face: Please file issues when you know something is wrong!)

Reply:
> (After spending hours trying to understand what was going on ☹️ Please file issues when you know something is wrong!)

Sorry, I'll pay more attention in the future.